### DISCUSSION TITLE
Encryption Hashing: trying to understand DTLS hashing algorithm

### DISCUSSION THREAD

**Discussion by storycoffee** [openthread/openthread#8290] [Q&A] [answered]
Hey, I am studying security of OpenThread and I am trying to better understand the DTLS handshake. I read this from an academic paper:

said paper can be found here: https://dl.acm.org/doi/pdf/10.1145/3507657.3528544

After a successful
completion of the Thread commissioning process, the joiner will
have received several network parameters including a master key.
According to Wireshark [ 54 ],2 given a 128-bit master key 𝐾Master,
a 32-bit sequence counter 𝑐, and an 8-bit key index 𝑖, if we first
compute a message 𝑚 as
𝑚 = ((𝑐 & 0x80) + ((𝑖 − 1) & 0x7f)) || 0x546872656164, (1)
then we can derive the 128-bit MAC key 𝐾MAC and the 128-bit MLE
key 𝐾MLE by computing
𝐾MLE || 𝐾MAC = HMAC_SHA256(𝐾Master, 𝑚), (2)
where & denotes the bitwise AND operation, || denotes the concate-
nation operation, and HMAC_SHA256(·, ·) denotes the keyed-hash
message authentication code (HMAC) function [ 33 ] that uses the
SHA-256 hash function [ 37]. In words, the MLE key corresponds
to the first 128 bits of the output of the HMAC-SHA256 function
that has been parameterized by the master key and the computed
message, while the last 128 bits correspond to the MAC key. An
equivalent derivation process can be observed in OpenThread’s
implementation of the Thread networking protocol.

It is difficult to understand fully what is happening in this algorithm especially with their use of the '||' symbol as concatenation. If anyone could help me out with this, show me where in the source code this happens, break it down to me, or even just point me in the right direction I would greatly appreciate it.

**Comment by jwhui** [comment, marked as answer]
The DTLS Handshake protocol is specified in [RFC 6347](https://datatracker.ietf.org/doc/html/rfc6347#section-4.2). OpenThread leverages [mbedTLS](https://www.trustedfirmware.org/projects/mbed-tls/) for the DTLS implementation.

The text you quoted discusses Thread's security material generation, which is specified in [Thread 1.1.1 Section 7.1](https://www.threadgroup.org/support#specifications) and implemented here: https://github.com/openthread/openthread/blob/bdb938297630578c1393d297b0ee0495a4a3ecdd/src/core/thread/key_manager.cpp#L290

**Reply by storycoffee** [reply]
Thank you for being so helpful
